networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/Eth2PeerSelectionStrategy.java

```diff
@@ -38,54 +47,149 @@
                 .reversed())
-        .map(network::createPeerAddress)
+        .flatMap(candidate -> checkCandidate(candidate, network).stream())
+        .limit(scoreBasedPeersToAdd)
+        .forEach(selectedPeers::add);
+    return selectedPeers;
+  }
+
+  private int getCurrentRandomlySelectedPeerCount(
+      final P2PNetwork<?> network, final PeerSources peerSources) {
+    return (int)
+        network
+            .streamPeers()
+            .filter(peer -> peerSources.getSource(peer.getId()) == RANDOMLY_SELECTED)
+            .count();
+  }
+
+  private Optional<PeerAddress> checkCandidate(
+      final DiscoveryPeer candidate, final P2PNetwork<?> network) {
+    return Optional.of(network.createPeerAddress(candidate))
         .filter(reputationManager::isConnectionInitiationAllowed)
-        .filter(peerAddress -> !network.isConnected(peerAddress))
-        .limit(maxAttempts)
-        .collect(toList());
+        .filter(peerAddress -> !network.isConnected(peerAddress));
   }
 
   @Override
   public List<Peer> selectPeersToDisconnect(
-      final P2PNetwork<?> network, final Predicate<Peer> canBeDisconnected) {
-    final int peersToDrop = targetPeerCountRange.getPeersToDrop(network.getPeerCount());
+      final P2PNetwork<?> network, final PeerSources peerSources) {
+
+    final Map<PeerSource, List<Peer>> peersBySource =
+        network
+            .streamPeers()
+            .collect(Collectors.groupingBy(peer -> peerSources.getSource(peer.getId())));
+
+    final List<Peer> randomlySelectedPeers =
+        peersBySource.getOrDefault(RANDOMLY_SELECTED, new ArrayList<>());
+    final int randomlySelectedPeerCount = randomlySelectedPeers.size();
+
+    final int currentPeerCount = network.getPeerCount();
+    final int peersToDrop = targetPeerCountRange.getPeersToDrop(currentPeerCount);
     if (peersToDrop == 0) {
       return emptyList();
     }
     final PeerScorer peerScorer = peerSubnetSubscriptionsFactory.create(network).createScorer();
-    return network
-        .streamPeers()
-        .filter(canBeDisconnected)
+    final int randomlySelectedPeersToDrop =
+        targetPeerCountRange.getRandomlySelectedPeersToDrop(
+            randomlySelectedPeerCount, currentPeerCount);
+    shuffler.shuffle(randomlySelectedPeers);
+    final List<Peer> randomlySelectedPeersBeingDropped =
+        randomlySelectedPeers.subList(
+            0, Math.min(randomlySelectedPeersToDrop, randomlySelectedPeers.size()));
+    // Peers from the randomly selected pool that have been chosen are now left to justify
+    // themselves based on their score alone.
+    randomlySelectedPeersBeingDropped.forEach(peer -> peerSources.forgetPeer(peer.getId()));
+    return Stream.concat(
+            randomlySelectedPeersBeingDropped.stream(),
+            peersBySource.getOrDefault(SELECTED_BY_SCORE, emptyList()).stream())
         .sorted(Comparator.comparing(peerScorer::scoreExistingPeer))
         .limit(peersToDrop)
         .collect(toList());
```

Comment:
Shouldn't we call `peerSources.forgetPeer(peer.getId()))` after `.limit(peersToDrop)` instead of line 182?
May be some of `randomlySelectedPeersBeingDropped` will not fall into the final list

Reply:
We want to move them out of the randomly selected pool which puts them into the default score based pool.  Whether they get disconnected or not then just depends on their score, but they aren't "protected" as randomly selected peers anymore.